### PR TITLE
[DO NOT SQUASH] Add a rock-check-residency pass to check occupancy/residency of a given kernel

### PR DESCRIPTION
**Pull request by sjw36**
This pass check various metrics collected during rock lowering and AMDGPU backend captured in the ISA metadata. 

Computes occupancy based on:
- Shared Buffer total size (rock lowering)
- VGPR allocation (amdgpu backend)
- SGPR allocation (amdgpu backend)

From these kernel requirements the rock-check-residency pass will compute blocks_per_cu occupancy within the arch target.

This PR also includes a new AttrName infrastructure to manage get/set utilities for NamedAttributes.
- Consolidated various rock NamedAttributes to use new infrastructure.

### CODE REVIEW

File: external/llvm-project/mlir/lib/Dialect/GPU/AmdDeviceLibsIncGen.py [hunk 1]
```diff
@@ -40,7 +40,7 @@ def generate(outputPath: Path, rocmPath: Path, libs: List[str]) -> None:
         print(f"{as_signed(byte):+4},", file=out, end=("\n" if i % 8 == 0 else " "))
       # Terminating null pointer needed for
       print("0x00};", file=out)
-    print("static constexpr std::initializer_list<std::pair<llvm::StringRef, llvm::StringRef>> allLibList = {", file=out)
+    print("static const std::initializer_list<std::pair<llvm::StringRef, llvm::StringRef>> allLibList = {", file=out)
```
Comment (krzysz00) [inline external/llvm-project/mlir/lib/Dialect/GPU/AmdDeviceLibsIncGen.py:43]: Why?

File: external/llvm-project/mlir/include/mlir/IR/AttrTypeBase.td [hunk 2]
```diff
@@ -262,6 +262,60 @@ class AttrDef<Dialect dialect, string name, list<Trait> traits = [],
                                  "::" # cppClassName # ">($_self)">;
 }
 
+// Define an Attr name for NamedAttribute's.
+// - Name is dialect-scoped when not-inherent.
+// - Utilities to has/get/set/lookup on an Operation
+class AttrNameDef<Dialect dialect, string name, string userName,
```
Comment (krzysz00) [inline external/llvm-project/mlir/include/mlir/IR/AttrTypeBase.td:268]: Maybe `DialectAttr`

File: external/llvm-project/mlir/include/mlir/IR/AttrTypeBase.td [hunk 3]
```diff
@@ -262,6 +262,60 @@ class AttrDef<Dialect dialect, string name, list<Trait> traits = [],
...
+    static bool hasOn(::mlir::Operation *op) {
+      return op->hasAttr(getName(op));
+    }
+    static ValueAttrType getOn(::mlir::Operation *op) {
+      return op->getAttrOfType<ValueAttrType>(getName(op));
+    }
+    static ValueAttrType lookupOn(::mlir::Operation *op) {
+      if (auto attr = getOn(op))
+        return attr;
+      std::optional<RegisteredOperationName> opInfo = op->getRegisteredInfo();
```
Comment (krzysz00) [inline external/llvm-project/mlir/include/mlir/IR/AttrTypeBase.td:294]: Strong object to this logic under this name.

Perhaps `findInContext` to have this recursive behavior

File: external/llvm-project/mlir/lib/Dialect/GPU/Transforms/SerializeToHsaco.cpp [hunk 4]
```diff
@@ -364,10 +364,55 @@ SerializeToHsacoPass::translateToLLVMIR(llvm::LLVMContext &llvmContext) {
   return ret;
 }
 
+/// Parse amdhsa specific metadata from the isa string.
+/// Note: the backend provides no other way to capture this data.
+/// TODO(sjw): add backend pass to collect structured metadata.
+static void applyMetadata(Operation *op, const std::string &isa) {
+  std::istringstream iss(isa);
```
Comment (krzysz00) [inline external/llvm-project/mlir/lib/Dialect/GPU/Transforms/SerializeToHsaco.cpp:371]: I think there're LLVM equivalents of this - go stare at LLVM's string handling?

File: mlir/include/mlir/Dialect/Rock/IR/RockAttrDefs.td [hunk 5]
```diff
@@ -173,8 +177,7 @@ def Rock_TransformAttr : Rock_Attr<"Transform"> {
 }
 
 /// TransformsAttr
-def Rock_TransformMapAttr : Rock_Attr<"TransformMap"> {
-    let mnemonic = "transform_map";
+def Rock_TransformMapAttr : Rock_Attr<"TransformMap", "transform_map"> {
```
Comment (krzysz00) [inline mlir/include/mlir/Dialect/Rock/IR/RockAttrDefs.td:180]: Strong object: these attributes aren't the sort of discardable/dialect-wide attributes your setup is meant for

File: mlir/include/mlir/Dialect/Rock/IR/RockAttrDefs.td [hunk 6]
```diff
@@ -339,7 +342,54 @@ def Rock_GemmFeatures : I32BitEnumAttr<"GemmFeatures",
...
+def Rock_NumCuAttr : Rock_I32Attr<"NumCu", "num_cu">;
+def Rock_BlockSizeAttr : Rock_I32Attr<"BlockSize", "block_size">;
+def Rock_GridSizeAttr : Rock_I32Attr<"GridSize", "grid_size">;
+def Rock_WaveSizeAttr : Rock_I32Attr<"WaveSize", "wave_size">;
+def Rock_SharedBufferSizeAttr : Rock_I32Attr<"SharedBufferSize", "shared_buffer_size">;
+
+def Rock_BlocksPerCUAttr : Rock_I32Attr<"BlocksPerCU", "blocks_per_cu">;
+
+def Rock_Is64BitIndexAttr : Rock_UnitAttr<"Is64BitIndex", "is_64_bit_index">;
+def Rock_HasGlobalSyncAttr : Rock_UnitAttr<"HasGlobalSync", "has_global_sync">;
```
Comment (krzysz00) [inline mlir/include/mlir/Dialect/Rock/IR/RockAttrDefs.td:393]: These, on the other hand, make sense

**Review by krzysz00 — CHANGES_REQUESTED** [review]
Ok, so this should be three PRs
1. The PR that adds this dialect interface thing (to which I have objections, see reviews comments). I personally claim that this PR should be developed upstream first and backported to here.
2. The PR that adds the metadata collection / global sync stuff
3. The PR that refactors how `rock` handles its annotating attributes

File: mlir/include/mlir/Dialect/Rock/IR/RockAttrDefs.td (hunk 5 above)
Comment (sjw36) [reply mlir/include/mlir/Dialect/Rock/IR/RockAttrDefs.td:180]: They aren't using the new infra. I just moved the mnemonic setting to Rock_Attr.

File: external/llvm-project/mlir/include/mlir/IR/AttrTypeBase.td (hunk 2 above)
Comment (sjw36) [reply external/llvm-project/mlir/include/mlir/IR/AttrTypeBase.td:268]: `AttrDef` already takes a Dialect.

File: external/llvm-project/mlir/lib/Dialect/GPU/AmdDeviceLibsIncGen.py (hunk 1 above)
Comment (sjw36) [reply external/llvm-project/mlir/lib/Dialect/GPU/AmdDeviceLibsIncGen.py:43]: Didn't compile with g++.

Comment (krzysz00) [reply external/llvm-project/mlir/lib/Dialect/GPU/AmdDeviceLibsIncGen.py:43]: It's unsupported for rocMLIR builds, and this script doesn't exist upstream

File: external/llvm-project/mlir/include/mlir/IR/AttrTypeBase.td (hunk 2 above)
Comment (krzysz00) [reply external/llvm-project/mlir/include/mlir/IR/AttrTypeBase.td:268]: Good point, but AttrNameDef still doesn't quite feel right - I'd argue for something like `ExternalAttrDef`

File: mlir/include/mlir/Dialect/Rock/IR/RockAttrDefs.td (hunk 5 above)
Comment (krzysz00) [reply mlir/include/mlir/Dialect/Rock/IR/RockAttrDefs.td:180]: Yeah, fair enough

File: external/llvm-project/mlir/include/mlir/IR/AttrTypeBase.td (hunk 3 above)
Comment (sjw36) [reply external/llvm-project/mlir/include/mlir/IR/AttrTypeBase.td:294]: Is it a **strong object** to the name or the logic?

Comment (krzysz00) [reply external/llvm-project/mlir/include/mlir/IR/AttrTypeBase.td:294]: The logic itself is something that could exist. It's a strong object to it existing under this name in this context.

`lookupOn(Operation *op)` should not be doing a tree-walk.

`findNearestTo(Operation *op)` or a similar name should do the traversal you have in `lookupOn(Operation *op)`

Comment (sjw36) [reply external/llvm-project/mlir/include/mlir/IR/AttrTypeBase.td:294]: This is a hierarchical scope lookup based on the philosophy of inheritance. Operations under a loop, function or module can inherit properties based on their `parent` scope. I don't think findNearestTo captures this idea.

**Comment by sjw36** [comment]
Superceded by https://github.com/ROCmSoftwarePlatform/rocMLIR/pull/1329

**Pull request closed without merging**